### PR TITLE
Heartbeat input plugin

### PR DESCRIPTION
**Pull request by varunkumta**
-- Add new heartbeat input plugin and lib
-- Update conf (Both operation and heartbeat will use the same output plugin)
-- Updates to oms_common to support get_os_name, get_os_version, get_fqdn, get_installed_date and get_agent_version. Also fix newline handling
-- Add tests for oms_common and heartbeat_lib

@Microsoft/omsagent-devs

### CODE REVIEW

File: source/code/plugins/heartbeat_lib.rb [hunk 1]
```diff
@@ -0,0 +1,64 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+module HeartbeatModule
+  class LoggingBase
+    def log_error(text)
+    end
+  end
+
+  class RuntimeError < LoggingBase
+    def log_error(text)
+      $log.info "RuntimeError: #{text}"
```
Comment (robbiezhang) [inline source/code/plugins/heartbeat_lib.rb]: should it log as error?

File: test/code/plugins/heartbeat_lib_test.rb [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    heartbeat_data_item = call_get_heartbeat(@tmp_os_conf_file.path, @tmp_agent_install_conf_file.path)
+    assert_equal(1, $log.logs.size, "No error in logs for malformed agent install conf file")
+    verify_invalid_heartbeat_data(heartbeat_data_item)
+  end
+
+  def test_heartbeat_valid_conf_files
+    File.write(@tmp_os_conf_file.path, "OSName=Ubuntu\nOSVersion=14.04\nOSFullName=Ubuntu 14.04 (x86_64)\nOSAlias=UniversalD\nOSManufacturer=Canonical Group Limited\n")
+    File.write(@tmp_agent_install_conf_file.path, "1.1.0-124 20160412 Release_Build\n2016-05-24T00:27:55.0Z\n")
+    heartbeat_data_item = call_get_heartbeat(@tmp_os_conf_file.path, @tmp_agent_install_conf_file.path)
+    assert(heartbeat_data_item.has_key?("Timestamp"))
```
Comment (niroyb) [inline test/code/plugins/heartbeat_lib_test.rb]: Also check the parsed value. You can do an assert_equal({...}, heartbeat_data_item)

File: test/code/plugins/heartbeat_lib_test.rb [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    heartbeat_data_item = call_get_heartbeat(@tmp_os_conf_file.path, @tmp_agent_install_conf_file.path)
+    verify_invalid_heartbeat_data(heartbeat_data_item)
+  end
+
+  def test_heartbeat_malformed_conf_files
+    $log = OMS::MockLog.new
+    File.write(@tmp_os_conf_file.path, "Malformed OS Name\nMalformed OS Version\n")
+    File.write(@tmp_agent_install_conf_file.path, "Agent Version\nInstalledDate\n")
+    heartbeat_data_item = call_get_heartbeat(@tmp_os_conf_file.path, @tmp_agent_install_conf_file.path)
+    assert_equal(1, $log.logs.size, "No error in logs for malformed agent install conf file")
```
Comment (niroyb) [inline test/code/plugins/heartbeat_lib_test.rb]: Consider printing the logs in the assert message.

**Comment by niroyb** [comment]
Consider removing the heartbeat code from `omsadmin.sh` if this replaces it.

File: source/code/plugins/heartbeat_lib.rb (hunk 1 above)
Comment (varunkumta) [inline source/code/plugins/heartbeat_lib.rb]: done

File: test/code/plugins/heartbeat_lib_test.rb (hunk 2 above)
Comment (varunkumta) [inline test/code/plugins/heartbeat_lib_test.rb]: checking assert_equal for parsed data (osname, osmajor and minor versions, installed date, and agent version)

**Comment by varunkumta** [comment]
the heartbeat code in omsadmin.sh is for our telemetry .. this is data that the user can search for .. I think they are separate for now

File: test/code/plugins/heartbeat_lib_test.rb (hunk 3 above)
Comment (varunkumta) [inline test/code/plugins/heartbeat_lib_test.rb]: done

**Comment by niroyb** [comment]
Signed off

**Pull request merged**